### PR TITLE
Fix ice behaviour in world

### LINKED ISSUE #4923: Proper packed ice behaviour
Cuberite Commit id: 23bca008317e125659d4b97f29cf13da81a7aecd

https://minecraft.gamepedia.com/Packed_Ice

### Expected behavior
When breaking packed ice no wather should spawn

### Actual behavior
it does

maybe there are some other things...

### PR DESCRIPTION
**Pull request by 12xx12**
This fixes some bahavior around ice blocks in the world.

- They now smelt properly
- The packed ice doesn't spawn anymore water now

fixes 4923 partially atm. got to check some speed bahaviour on packed ice

### CODE REVIEW

File: src/Blocks/BlockIce.h [hunk 1]
```diff
@@ -31,6 +31,44 @@ class cBlockIceHandler :
...
+		cWorldInterface & a_WorldInterface,
+		cBlockPluginInterface & a_PluginInterface,
+		cChunk & a_Chunk,
+		const Vector3i a_RelPos
+	) const override
+	{
+		if (a_WorldInterface.GetDimension() == dimNether)
+		{
+			a_Chunk.SetBlock(a_RelPos, E_BLOCK_AIR, 0);
+		}
```
Comment (tigerw) [inline src/Blocks/BlockIce.h:45]: early return here perhaps?

File: src/Blocks/BlockIce.h [hunk 2]
```diff
@@ -31,6 +31,44 @@ class cBlockIceHandler :
...
+		{
+			Pos = StartAbsPos + Coords[i];
+			Chunk = a_Chunk.GetNeighborChunk(Pos.x, Pos.z);
+			Smelt |= Chunk->GetBlockLight(Chunk->AbsoluteToRelative(Pos)) > 11;
+		}
+
+		if (Smelt)
+		{
+			a_Chunk.SetBlock(a_RelPos, E_BLOCK_STATIONARY_WATER, 0);
+		}
```
Comment (tigerw) [inline src/Blocks/BlockIce.h:69]: I would suggest a couple changes: use std::array, keep coordinates relative, use ranged for loop, queue lighting, and early return; something like:

```cpp
std::array<Vector3i, 7> Adjacents = {
  // ...
};

for (const auto Offset : Adjacents)
{
  const auto Position = a_RelPos + Offset;
  const auto Chunk = a_Chunk.GetRelNeighborChunk();

  if (!Chunk->IsLightValid())
  {
    Chunk->GetWorld()->QueueLightChunk();
    return;
  }

  if (Chunk->GetBlockLight(Position) > 11)
  {
    Chunk->SetBlock(Position, WATER);
    return;
  }
}

File: src/Blocks/BlockIce.h [hunk 3]
```diff
@@ -31,6 +31,44 @@ class cBlockIceHandler :
...
+		// Artificial light on any of the surrounding block > 11 leads to melting ice
+		bool Smelt = false;
+		Vector3i Pos;
+		cChunk *Chunk;
+
+		Vector3i Coords[7] =
+			{{1, 0, 0}, {-1, 0, 0},
+			{0, 1, 0}, {0, -1, 0},
+			{0, 0, 1}, {0, 0, -1},
+			{0, 0, 0}};
```
Comment (tigerw) [inline src/Blocks/BlockIce.h:55]: by the way is 0,0,0 needed?

Comment (12xx12) [reply src/Blocks/BlockIce.h:55]: It did show some strange behaviour when I didn't add it. The wiki says only surrounding blocks but that didn't melt the blocks like in vanilla

File: src/Blocks/BlockPackedIce.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+private:
+
+	virtual cItems ConvertToPickups(NIBBLETYPE a_BlockMeta, cBlockEntity * a_BlockEntity, const cEntity * a_Digger, const cItem * a_Tool) const override
+	{
+		// Only drop self when using silk-touch:
+		if (ToolHasSilkTouch(a_Tool))
+		{
+			return cItem(m_BlockType);
+		}
+		else
```
Comment (tigerw) [inline src/Blocks/BlockPackedIce.h:28]: Since there's a return already, this else can be removed, I suppose.

File: src/Blocks/BlockPackedIce.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+#pragma once
+
+#include "BlockHandler.h"
+
+
+
+
+
+class cBlockPackedIceHandler :
+		public cBlockHandler
```
Comment (tigerw) [inline src/Blocks/BlockPackedIce.h:11]: `tabtab`! consider `tab` instead

File: src/Blocks/BlockPackedIce.h (hunk 4 above)
Comment (12xx12) [reply src/Blocks/BlockPackedIce.h:28]: But the else is run if the tool was't with Silk touch

File: src/Blocks/BlockIce.h (hunk 3 above)
Comment (tigerw) [reply src/Blocks/BlockIce.h:55]: could be the lighting being invalid perhaps? does sound strange.

Comment (12xx12) [reply src/Blocks/BlockIce.h:55]: Maybe. It just melted if the light source block was at the same y coords. Doing some more research tomorrow

File: src/Blocks/BlockPackedIce.h (hunk 4 above)
Comment (12xx12) [reply src/Blocks/BlockPackedIce.h:28]: No i got it

File: src/Blocks/BlockIce.h [hunk 6]
```diff
@@ -25,9 +25,47 @@ class cBlockIceHandler :
...
+		const std::array<Vector3i, 7> Offsets =
+			{Vector3i(1, 0, 0), Vector3i(-1, 0, 0),
+			Vector3i(0, 1, 0), Vector3i(0, -1, 0),
+			Vector3i(0, 0, 1), Vector3i(0, 0, -1)};
+
+		for (const auto & Offset : Offsets)
 		{
-			return {};
+			auto Position = a_RelPos + Offset;
+			const auto Chunk = a_Chunk.GetRelNeighborChunk(Position.x, Position.z);
```
Comment (tigerw) [inline src/Blocks/BlockIce.h:54]: You can GetRelNeighborChunkAdjustCoords to make Position relative to the returned chunk

Comment (12xx12) [reply src/Blocks/BlockIce.h:54]: Oh so that's what that method does :D

**Comment by 12xx12** [comment]
The smelting across chunk borders works now. Checked at the corner where 4 chunks met

**Comment by tigerw** [comment]
Uno momento por favor

**Comment by 12xx12** [comment]
Yup, changes seem appropriate

File: src/Blocks/BlockIce.h [hunk 7]
```diff
@@ -25,9 +25,54 @@ class cBlockIceHandler :
 		{
 			return cItem(m_BlockType);
 		}
```
Comment (12xx12) [inline src/Blocks/BlockIce.h:27]: Now I am confused

Comment (tigerw) [reply src/Blocks/BlockIce.h:27]: diff looks weird?

Comment (12xx12) [reply src/Blocks/BlockIce.h:27]: I ment to mark then line with ````{{``` why those are needed

Comment (tigerw) [reply src/Blocks/BlockIce.h:27]: The std::array {{?

I think the first { is to indicate initialisation of std::array, the second { indicates you want to aggregate-initialise the C-array the std::array contains, and the remaining {x,y,z} initialises the contained Vector3i

Comment (12xx12) [reply src/Blocks/BlockIce.h:27]: ok - thank you

**Review by tigerw — APPROVED** [review]

**Pull request merged**